webapp/src/main/java/com/box/l10n/mojito/service/openai/OpenAIService.java

```diff
@@ -0,0 +1,301 @@
+  }
+
+  @Timed("OpenAIService.checkString")
+  private List<OpenAICheckResult> checkString(
+      AssetExtractorTextUnit textUnit, List<AIPrompt> prompts, long repositoryId) {
+    try {
+      List<OpenAICheckResult> results = new ArrayList<>();
+      String sourceString = textUnit.getSource();
+      String comment = textUnit.getComments();
+      String[] nameSplit = textUnit.getName().split(" --- ");
+
+      if (!prompts.isEmpty()) {
+        executePromptChecks(
+            textUnit, prompts, repositoryId, sourceString, comment, nameSplit, results);
+      } else {
+        logger.warn("No prompts found for repository id: {}", repositoryId);
+        OpenAICheckResult result = new OpenAICheckResult();
+        result.setSuccess(true);
+        result.setSuggestedFix(
+            "No prompts found for repository id: " + repositoryId + ", skipping check.");
+        results.add(result);
+      }
+
+      return results;
+    } catch (JsonProcessingException e) {
+      logger.error("Error processing JSON response from OpenAI", e);
+      throw new OpenAIException("Error processing JSON response from OpenAI", e);
+    }
+  }
+
+  private void executePromptChecks(
+      AssetExtractorTextUnit textUnit,
+      List<AIPrompt> prompts,
+      long repositoryId,
+      String sourceString,
+      String comment,
+      String[] nameSplit,
+      List<OpenAICheckResult> results)
+      throws JsonProcessingException {
+    for (AIPrompt prompt : prompts) {
+      if ((!Strings.isNullOrEmpty(prompt.getSystemPrompt())
+              && !prompt.getSystemPrompt().contains(SOURCE_STRING_PLACEHOLDER))
+          && (!Strings.isNullOrEmpty(prompt.getUserPrompt())
+              && !prompt.getUserPrompt().contains(SOURCE_STRING_PLACEHOLDER))) {
+        logger.error(
+            "Source string placeholder is missing in both system and user prompts for prompt id {}. Skipping check.",
+            prompt.getId());
+        continue;
+      }
+      String systemPrompt = getFormattedPrompt(prompt.getSystemPrompt(), sourceString, comment);
+
+      String userPrompt = getFormattedPrompt(prompt.getUserPrompt(), sourceString, comment);
+
+      if (nameSplit.length > 1
+          && (systemPrompt.contains(CONTEXT_STRING_PLACEHOLDER)
+              || userPrompt.contains(CONTEXT_STRING_PLACEHOLDER))) {
+        systemPrompt = systemPrompt.replace(CONTEXT_STRING_PLACEHOLDER, nameSplit[1]);
+        userPrompt = userPrompt.replace(CONTEXT_STRING_PLACEHOLDER, nameSplit[1]);
+      }
+
+      OpenAIClient.ChatCompletionsRequest chatCompletionsRequest =
+          buildChatCompletionsRequest(prompt, systemPrompt, userPrompt);
+
+      OpenAIClient.ChatCompletionsResponse chatCompletionsResponse =
+          openAIClient.getChatCompletions(chatCompletionsRequest).join();
+      persistCheckResult(textUnit, repositoryId, prompt, chatCompletionsResponse);
+      OpenAICheckResult result =
+          objectMapper.readValue(
```

Comment:
From prior experience, I don't always expect the model to come back with a correctly formatted response, i.e.: it might provide an incorrect json back, use a different return key than the expect one for one of the fields, have extra unexpected fields, not follow the instruction given etc. 

If processing a list of 100 source strings in a batch, it would be reasonable to expect that at least one would suffer from that issue (and it might be a different one each time). That means that the larger the set of changes submitted is - the higher likelyhood of the entire check failing due one of these mis-formatting issues.

Have you seen this in testing by any chance?

Some approaches to fixing this:
- plain retrying until we get a parse-able response (not guaranteed to succeed) 
- increasing num_completions on retries to be able to go through multiple answers and pick the first "good" one
- incrementally increasing the temperature on each failure when retrying for the same input & prompt combination

Any thoughts on this?

Reply:
Testing that I've done so far has been pretty basic prompts to test the E2E flow. I haven't yet observed the above but I imagine that's due to the limited prompt testing that's been done up to this point. Further testing is required to establish the prompt that will be used for the checks.

I can catch a JSON processing exception within the loop to avoid all checks being returned as an error, from your experience what was the most 'helpful' approach to getting parseable output back?

Reply:
Adding:
`        response_format={"type": "json_object"},`
to the request might help significantly is the desired output is json - though will still sometimes result in incorrect responses (wrong keys, not correctly escaped json etc.).